### PR TITLE
Introduce natural sorting

### PR DESCRIPTION
**Pull request by garster**
No doubt that DB ordering is way faster than sorting in code, but to me this is big deal. It would be great to get some speed numbers, but also big-picture, speed improvements could be done in other places too.

I don't know what Laravel uses under the hood for sorting (I can't find the source code), but PHP says it uses Quicksort which works well on a list that is already well sorted.

So my suggested sledgehammer here is to add a sortBy onto just about every orderBy. The DB will sort well then the sortBy will have another pass.

So far it is worked will on a album with 150 pics, I get the API response back in about 200ms. I have not done a comparison.

This PR is related to #321

### CODE REVIEW

File: app/ModelFunctions/AlbumFunctions.php [hunk 1]
```diff
@@ -543,7 +545,9 @@ public function getToplevelAlbums()
 			if ($id == 0) {
 				$return['albums'] = Album::where('owner_id', '=', 0)
 					->where('parent_id', '=', null)
-					->orderBy(Configs::get_value('sorting_Albums_col'), Configs::get_value('sorting_Albums_order'))->get();
+					->orderBy(Configs::get_value('sorting_Albums_col'), Configs::get_value('sorting_Albums_order'))
+					->get()
+					->sortBy(Configs::get_value('sorting_Albums_col'), SORT_NATURAL, (Configs::get_value('sorting_Albums_order') === 'DESC'));
```
Comment (kamil4) [inline app/ModelFunctions/AlbumFunctions.php]: So, given your follow-up timings from #321, I wonder if we should be doing orderBy here and in other such places at all... I think we'll need to try three versions:

* plain `orderBy` (current master)
* `orderBy` plus `sortBy` (your PR)
* just `sortBy`

I'll see if I can work on that tonight...

File: app/ModelFunctions/AlbumFunctions.php [hunk 2]
```diff
@@ -136,7 +136,9 @@ public function gen_thumbs($return, $album_list)
 		$photos = Photo::whereIn('album_id', $album_list)
 			->orderBy('star', 'DESC')
 			->orderBy(Configs::get_value('sorting_Photos_col'), Configs::get_value('sorting_Photos_order'))
-			->limit(3)->get();
+			->limit(3)
+			->get()
+			->sortBy(Configs::get_value('sorting_Photos_col'), SORT_NATURAL, (Configs::get_value('sorting_Photos_order') === 'DESC'));
```
[the remark this answers is not in the document]
Comment (garster) [reply app/ModelFunctions/AlbumFunctions.php]: I did not test photos, and the sortBy must be after the ->get() so it will not work well in this situation. I think the sortBy should be removed here. I  will modify the code.

**Comment by garster** [comment]
> I'm also surprised by the lack of changes to `set_order` in `Photo.php`. Did you verify that your branch sorts photos naturally in addition to albums?

I missed this, the trick is the sortBy has to come after the get() call. I traced the `set_order` function thru some others and found two places it is called. I made the change. I also saw it in DemoController but I did not make the change there.

File: app/ModelFunctions/AlbumFunctions.php (hunk 1 above)
Comment (garster) [reply app/ModelFunctions/AlbumFunctions.php]: Removing the OrderBys would simplify the code, I just did not want to get too crazy as an outside contributor.

**Comment by kamil4** [comment]
Yes, I made a similar change in my tree but I didn't get to test it yet as I spent most of the last evening creating an album with >10,000 pictures to have something to test the performance of sorting with.

**Comment by kamil4** [comment]
Sorry for the lack of updates. The tests I ran revealed some missing corner cases that were difficult to fix without large performance losses but I think I solved the major ones. I just need to clean up my code before pushing a commit on top of what's here.

File: app/ModelFunctions/AlbumFunctions.php [hunk 3]
```diff
@@ -517,11 +548,11 @@ public function setContentsOwner($albumID, int $ownerId)
...
 	 * their content).
 	 *
-	 * @param $parentAlbum
 	 * @param array $return
-	 * @param bool  $includePassProtected
+	 * @param $parentAlbum
+	 * @param bool $includePassProtected
 	 */
-	public function get_sub_albums($parentAlbum, array &$return, $includePassProtected = false)
+	public function get_sub_albums(array &$return, $parentAlbum, $includePassProtected = false)
```
Comment (ildyria) [inline app/ModelFunctions/AlbumFunctions.php]: what is the type of `$parentAlbum` ?

File: app/ModelFunctions/AlbumFunctions.php [hunk 4]
```diff
@@ -556,52 +587,60 @@ public function getToplevelAlbums()
 			'shared_albums' => null,
 		);
 
+		$sortingCol = Configs::get_value('sorting_Albums_col');
+		$sortingOrder = Configs::get_value('sorting_Albums_order');
+
+		$customSort = function ($query) use ($sortingCol, $sortingOrder) {
```
Comment (ildyria) [inline app/ModelFunctions/AlbumFunctions.php]: you really like functors. ^^
